jlm/rvsdg/view.hpp

```diff
@@ -16,18 +16,46 @@ namespace jlm::rvsdg
 
 class region;
 
+/**
+ * Prints the given rvsdg region to a string,
+ * through recursive traversal of nodes and subregions.
+ * All outputs are given unique names to show dataflow in the graph.
+ * @param region the region to be printed
+ * @return the string describing the region.
+ * @see view(region, map)
+ */
+std::string
+view(const jlm::rvsdg::region * region);
+
+/**
+ * Prints the given rvsdg region to a string, and exposes the unique name given to each output.
+ * @param region the region to be printed
+ * @param map the map where each rvsdg::output is mapped to its unique name.
+ * Outputs without names will have a name added.
+ * @return the string describing the region.
+ */
+std::string
+view(const jlm::rvsdg::region * region, std::unordered_map<const output *, std::string> & map);
+
+/**
+ * Recursively traverses and prints the given rvsdg region to the given file.
+ * @param region the region to be printed.
+ * @param out the file to be written to.
+ */
 void
 view(const jlm::rvsdg::region * region, FILE * out);
 
-static inline void
+/**
+ * Recursively traverses and prints the root region of the given rvsdg graph to the given file.
+ * @param graph the rvsdg graph to be printed.
+ * @param out the file to be written to.
+ */
+inline void
```

Comment:
I think the `static` needs to stay, otherwise the linker will give you an error when trying to link two object files where `view.hpp` was included. The `include` can go as far as I can see.

Reply:
That's what inline means, two object files can have their own definition, and they can safely be merged at link time, resulting in only one being left in the final binary. Static means each object file gets its own implementation.

Having both (top-level) static and inline will never make sense, and static "wins". There are also a bunch of rules about when inline is implicit, e.g. when you define method bodies inside a class/struct definition, or when you define a function with template parameters.

Reply:
You are absolutely right. I must have been asleep when I wrote it.